### PR TITLE
client/multi: Refactor MM Balance Handling

### PR DESCRIPTION
**Pull request by martonp**
This PR makes market making dependent on the `WalletHistorian` interface, which will need to be implemented for all wallets. Currently only BTC/clone SPV wallets and ETH/Polygon wallets can be used.

```
client/{asset,core,rpcserver}: Add WalletTransaction

- This adds a WalletTransaction function to the WalletHistorian interface.
It returns information about a transaction which the wallet made, or one
in which the wallet received funds. The TransactionConfirmations function
is removed, as it will no longer be needed once all wallets support
WalletTransaction.

- A `PartOfActiveBalance` field is added to the `WalletTransaction` type.
This is required for market making to know when it is safe to consider the
funds from an incoming transaction available for use.

- A `ParentAssetLockedAmt` field is added to the `core.Order` type. This is
needed in order to know the amount of token's fee asset is locked for an
order.
```

```
client/mm: Refactor bot balance tracking

- This diff updates the balance tracking for market making bots. Previously,
there was a fragile diff based accounting technique. Now, on each bot
action, including trades, withdrawals, and deposits, a pending action is
stored. This pending action is updated until it is complete, at which time
it is removed, and the "base" balance of the bot is updated. The bot
balances are calculated by adding the base balances of the bot to the
effects of all pending actions.

- The `wrappedCore` and `wrappedCEX` types are removed, and their functionalities
are combined into a `unifiedExchangeAdaptor`. This is useful because
pending deposits and pending withdrawals have balance effects on both
CEX and DEX balances. Each `unifiedExchangeAdaptor` listens to a core
notification stream, and the `MarketMaker` type no longer needs to. In
a future refactor, the bots themselves will no longer listen to the
core notification functionality, and all common functionalities can be
moved to the `unifiedExchangeAdaptor`.

- Previously, the fee assets of tokens were not taken into account, but
now they are.

- The Binance library now uses a different endpoint, `/api/v3/balances`,
to get balances. This is supported by the Binance testnet api, so
testbinance is no longer required for balances.
```

### CODE REVIEW

**Review by JoeGruffins — APPROVED** [review]
Just running the basic bot without arb. Do not need to fix now but see these errors with cancel orders pop up:
```
2024-01-02 05:17:00.021 [ERR] MM[MarketMaker-127.0.0.1:17273-42-0]: Error canceling order 9d70534d61d934e097e00c37ce8b0959b5b3b2cded5947b6c4bbf049fc549149: order 9d70534d61d934e097e00c37ce8b0959b5b3b2cded5947b6c4bbf049fc549149 - only one cancel order can be submitted per order per epoch. still waiting on cancel order 068a0d0b27a7a8c666f68f7c7da84f54374034fb40fed3b2dda6e1dcdeec284c to match
```
```
2024-01-02 05:44:00.036 [ERR] MM[MarketMaker-127.0.0.1:17273-42-0]: Error canceling order 4b9b95ee7d8b02899a220d3a5547be362ab35c6706f7543b319ea3cae4f4ea8f: order 4b9b95ee7d8b02899a220d3a5547be362ab35c6706f7543b319ea3cae4f4ea8f not cancellable in status canceled
```

File: client/mm/exchange_adaptor.go [hunk 1]
```diff
@@ -0,0 +1,1347 @@
...
+	fee         uint64
+	amtCredited uint64
+	// feeConfirmed is relevant assets with dynamic fees where the fee is
+	// not known until the tx is mined.
+	feeConfirmed bool
+	cexConfirmed bool
+	// cexConfirmed == true && success == false means the CEX did not credit the
+	// deposit for some reason.
+	success bool
+}
```
Comment (buck54321) [inline client/mm/exchange_adaptor.go:94]: Please separate the `mtx` and group with its protected fields

File: client/mm/exchange_adaptor_test.go [hunk 2]
```diff
@@ -0,0 +1,2943 @@
...
+							Status:    order.OrderStatusBooked,
+							ID:        orderIDs[1][:],
+							Matches: []*core.Match{
+								{
+									Swap: &core.Coin{
+										ID: walletTxIDs[1][:],
+									},
+								},
+							},
+						},
```
Comment (buck54321) [inline client/mm/exchange_adaptor_test.go:658]: Please try to identify places where you can use variables and helper function for populating test data. It's hard to identify the parameters being tested.

File: client/mm/libxc/binance.go [hunk 3]
```diff
@@ -425,17 +426,43 @@ func newBinance(apiKey, secretKey string, log dex.Logger, net dex.Network, binan
...
+// balances map.
+func (bnc *binance) setBalances(ctx context.Context) error {
+	var resp struct {
+		Balances []struct {
+			Asset  string  `json:"asset"`
+			Free   float64 `json:"free,string"`
+			Locked float64 `json:"locked,string"`
+		} `json:"balances"`
+	}
+	err := bnc.getAPI(ctx, "/api/v3/account", nil, true, true, &resp)
```
Comment (buck54321) [inline client/mm/libxc/binance.go:439]: Hmm. I made some changes in #2615 that might be relevant to these changes too. I think we should just expect to be connected to configured cexes, or connect on-the-fly when we need info.

Comment (martonp) [reply client/mm/libxc/binance.go:439]: This change just updates the API used to get the balances. I think your changes over there are still good.

> I think we should just expect to be connected to configured cexes, or connect on-the-fly when we need info.

I think the use should make some action before connecting the the CEX. It shouldn't automatically connect at start-up even if the user doesn't intend to use market making at all. They may want to set up some configurations before connecting.

File: client/mm/exchange_adaptor.go [hunk 4]
```diff
@@ -0,0 +1,1347 @@
...
+		availableDiff[fromFeeAsset] -= int64(o.ParentAssetLockedAmt + o.RefundLockedAmt)
+		availableDiff[toFeeAsset] -= int64(o.RedeemLockedAmt)
+		if o.FeesPaid != nil {
+			availableDiff[fromFeeAsset] -= int64(o.FeesPaid.Funding)
+		}
+
+		locked := map[uint32]uint64{}
+		locked[fromAsset] += o.LockedAmt
+		locked[fromFeeAsset] += o.ParentAssetLockedAmt + o.RefundLockedAmt
+		locked[toFeeAsset] += o.RedeemLockedAmt
```
[the remark this answers is not in the document]
Comment (martonp) [reply client/mm/exchange_adaptor.go:401]: They are very similar when the trade is first made, but you can see in `updatePendingDEXOrder`, they will be different later on.

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/cmd/testbinance/main.go [hunk 5]
```diff
@@ -194,8 +162,695 @@ type fakeBinance struct {
...
-	ci := f.coinInfo()
-	writeJSONWithStatus(w, ci, http.StatusOK)
+	// Returning configs for eth, btc, ltc, bch, zec, usdc, matic.
+	// Balances do not use a sapi endpoint, so they do not need to be handled
+	// here.
+	resp := `[
+		{
+			"coin": "MATIC",
+			"depositAllEnable": true,
+			"free": "0",
```
[the remark this answers is not in the document]
Comment (martonp) [reply client/cmd/testbinance/main.go:172]: No, we're not using this request to get the balances anymore. Using `/api/v3/account` which is supported by the Binance testnet.

File: client/mm/exchange_adaptor.go [hunk 6]
```diff
@@ -0,0 +1,1341 @@
...
+		feeAsset = token.ParentID
+	} else {
+		feeAsset = deposit.assetID
+	}
+	if u.baseDexBalances[feeAsset] < deposit.fee {
+		u.log.Errorf("%s balance on dex %d < deposit fee %d",
+			dex.BipIDSymbol(feeAsset), u.baseDexBalances[feeAsset], deposit.fee)
+		u.baseDexBalances[feeAsset] = 0
+	} else {
+		u.baseDexBalances[feeAsset] -= deposit.fee
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/mm/exchange_adaptor.go:677]: Let's just roll with this for now. Will revisit later.

**Review by buck54321 — APPROVED** [review]

**Comment by buck54321** [comment]
Rebase please.

**Pull request merged**